xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java

```diff
@@ -0,0 +1,299 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.envoyproxy.envoy.api.v2.auth.DownstreamTlsContext;
+import io.envoyproxy.envoy.api.v2.core.Node;
+import io.grpc.Internal;
+import io.grpc.InternalLogId;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.internal.ExponentialBackoffPolicy;
+import io.grpc.internal.GrpcUtil;
+import io.grpc.internal.ObjectPool;
+import io.grpc.internal.SharedResourceHolder;
+import io.grpc.xds.EnvoyServerProtoData.CidrRange;
+import io.grpc.xds.EnvoyServerProtoData.FilterChain;
+import io.grpc.xds.EnvoyServerProtoData.FilterChainMatch;
+import io.netty.channel.Channel;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.epoll.Epoll;
+import io.netty.channel.epoll.EpollEventLoopGroup;
+import io.netty.util.concurrent.DefaultThreadFactory;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * Serves as a wrapper for {@link XdsClientImpl} used on the server side by {@link
+ * io.grpc.xds.internal.sds.XdsServerBuilder}.
+ */
+@Internal
+public final class XdsClientWrapperForServerSds {
+  private static final Logger logger =
+      Logger.getLogger(XdsClientWrapperForServerSds.class.getName());
+
+  private static final EventLoopGroupResource eventLoopGroupResource =
+      Epoll.isAvailable() ? new EventLoopGroupResource("GrpcServerXdsClient") : null;
+  private final InternalLogId logId;
+  // Must be accessed from the syncContext
+  private boolean panicMode;
+  final SynchronizationContext syncContext =
+      new SynchronizationContext(
+        new Thread.UncaughtExceptionHandler() {
+          @Override
+          public void uncaughtException(Thread t, Throwable e) {
+            logger.log(
+                Level.SEVERE,
+                  "[" + getLogId() + "] Uncaught exception in the SynchronizationContext. Panic!",
+                  e);
+            panic(e);
+          }
+        });
+
+  private EnvoyServerProtoData.Listener curListener;
+  @Nullable private ObjectPool<XdsClient> xdsClientPool;
+  @Nullable private XdsClient xdsClient;
+
+  /** Wraps the XdsClientImpl for use by SdsProtocolNegotiators. */
+  public XdsClientWrapperForServerSds(int port, Bootstrapper bootstrapper) {
+    this.logId = InternalLogId.allocate("GrpcServer", Integer.toString(port));
+    Bootstrapper.BootstrapInfo bootstrapInfo = null;
+    try {
+      bootstrapInfo = bootstrapper.readBootstrap();
+    } catch (IOException e) {
+      logger.log(Level.SEVERE, "Error from readBootstrap", e);
```

Comment:
The exception goes into nowhere except a log message. And this is a constructor. This pattern is bad.

Same for other places, error handling/propagation in this class is in appropriate.

Reply:
Good point. At this point there are no "listeners" like you have the `NameResolver.Listener2`  to propagate this error to. Will wrap this into a runtime exception.